File: 79. Word Search/79. Word Search.cs
public class Solution {
    public bool Exist(char[][] board, string word) {
        int m = board.Length, n = board[0].Length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (DFS(board, i, j, word, 0)) return true;
            }
        }
        return false;
    }
    
    private bool DFS(char[][] board, int i, int j, string word, int k) {
        if (k == word.Length) return true;
        if (i < 0 || i >= board.Length || j < 0 || j >= board[0].Length || board[i][j] != word[k]) return false;
        
        char temp = board[i][j];
        board[i][j] = '#';
        bool found = DFS(board, i+1, j, word, k+1) ||
                     DFS(board, i-1, j, word, k+1) ||
                     DFS(board, i, j+1, word, k+1) ||
                     DFS(board, i, j-1, word, k+1);
        board[i][j] = temp;
        return found;
    }
}

File: 79. Word Search/79. Word Search.go
func exist(board [][]byte, word string) bool {
    for i := 0; i < len(board); i++ {
        for j := 0; j < len(board[0]); j++ {
            if dfs(board, i, j, word, 0) {
                return true
            }
        }
    }
    return false
}

func dfs(board [][]byte, i, j int, word string, k int) bool {
    if k == len(word) {
        return true
    }
    if i < 0 || i >= len(board) || j < 0 || j >= len(board[0]) || board[i][j] != word[k] {
        return false
    }
    
    temp := board[i][j]
    board[i][j] = '#'
    found := dfs(board, i+1, j, word, k+1) ||
             dfs(board, i-1, j, word, k+1) ||
             dfs(board, i, j+1, word, k+1) ||
             dfs(board, i, j-1, word, k+1)
    board[i][j] = temp
    return found
}